### ISSUE TITLE
Is "jaxlie.manifold.value_and_grad" the same as defined in https://arxiv.org/pdf/2103.12032.pdf

### ISSUE THREAD

**Issue by JonathonLuiten** [brentyi/jaxlie#8]
I am wondering if the 'tangent-space backpropagation' implemented here in "jaxlie.manifold.value_and_grad" is the same as that defined in the paper "Tangent Space Backpropagation for 3D Transformation Groups" (https://arxiv.org/pdf/2103.12032.pdf), with code at https://github.com/princeton-vl/lietorch, but implemented in jax instead of pytorch?

If there is any differences between the two approaches it would be nice to know what they are? 

If it is irrelevant to me if I use pytorch or jax, then is there any reason to use jaxlie vs lietorch?

Thanks for the answer. :)

**Comment by brentyi** [comment]
Hi!

> If there is any differences between the two approaches it would be nice to know what they are?

If you just need a black box for working with poses and/or pose optimization, they should be basically the same.

Under-the-hood, there are a couple of differences:

1. Minor convention stuff. Based on the paper it seems like `lietorch` uses a left-action parameterization for local parameterizations, eg for some loss function $f$ and transform $T$, $\nabla_{\xi}\ f(e^{\widehat{\xi}} T)$ is computed. `jaxlie` uses a right-action $\nabla_{\xi}\ f(T e^{\widehat{\xi}})$. `lietorch` uses xyzw for quaternions, `jaxlie` uses wxyz.

2. My understanding of `lietorch`/the very cool tangent-space backprop paper is that they backprop not only with respect to tangent spaces, but directly "on" them in the sense that the computations used to compute the gradients are invariant to the parameterization/embedding space used by the transformation. This requires a little fussing with PyTorch internals, but is really elegant mathematically. In contrast, `jaxlie` just blindly uses autodiff to backprop *with respect to* tangent spaces, which will (invisibly) require intermediate gradients with respect to the embedding space (eg quaternions for `SO(3)`).
    - Facebook's [theseus](https://arxiv.org/abs/2207.09442) also implements a similar approach as `jaxlie`, and spends a paragraph discussing implementation + difference from `lietorch` in their [paper](https://arxiv.org/abs/2207.09442).

> If it is irrelevant to me if I use pytorch or jax, then is there any reason to use jaxlie vs lietorch?

I think it should rarely matter.

The advantages of `lietorch` stem from it being smarter: there are fewer floating point operations, so I'd expect it to be faster and more precise. This matters especially for applications bottlenecked by backprop through transform operations. (eg rotation averaging)

The advantages of `jaxlie` are mostly associated with it being dumber and/or written using vanilla JAX primitives:
- Support for higher order derivatives. (maybe useful for metalearning, or any operation that requires backprop through a Jacobian)
- Forward-mode autodiff, this is nice for tall Jacobians.
- Interface makes fewer assumptions and is more general; it'd be easy for example to compare a Riemannian optimization approach against a projected one.
- TPU support.
- Type annotations for those who care about that kind of thing :-)

Let me know if anything's still unclear!

**Comment by JonathonLuiten** [comment]
Awesome!

That answers that question, but it does raise more questions which is:

What is exactly is the difference between the 3 methods presented for SE3 optimization in the example code here: https://github.com/brentyi/jaxlie/blob/master/examples/se3_optimization.py

How is *backprop wrt to tangent space* (method 1) different from differentiation directly on either an axis angle representation (method 3) or directly on quaternions (I'm guessing it's possible to create a method that does this similar to method 3), or by projecting the updated matrix elements back onto the SE3 manifold (which I assume does SVD like this paper: https://arxiv.org/abs/2006.14616) (method 2).

I'm assuming the practical difference between these methods might be about computing derivatives around certain points in the SE3 space, e.g. like around (axis-angle = [0,0,0,0,0,0]). Does your method (method 1) solve such issues, that I think are present in method 3??

**Comment by brentyi** [comment]
> What is exactly is the difference between the 3 methods presented for SE3 optimization

Sure, so  let's say we have a loss function $f(T)$, where $T$ is some transformation matrix.

For SE3 in `jaxlie` we typically parameterize $T$ with a length-7 vector (quaternion + translation). The problem is that we can't just naively run gradient descent on these parameters, right? Because there's some manifold structure that needs to be respected to keep them valid.

All 3 of the methods are just trying to solving this, in order of increasing complexity:

**Method 2 (projected gradient descent)** in the example is probably the simplest. We compute backprop normally wrt the parameters to get a gradient of size `(7,)`, apply the ADAM scaling rules, then add the update. This may produce an invalid rotation, but we can project after each step to fix that.

As you mentioned SVD can be used for projection/orthogonalization of rotation matrices, but in `jaxlie` we use unit quaternions so projection back to the manifold can be done via simple normalization.

**Method 3 (exponential coordinates)** proposes to reparameterize our transformation using the Lie algebra. Instead of tracking the global parameterization of $T$, we instead only store $\xi = \text{logm}(T^\vee) \in \mathbb{R}^6$. $T$ can then be recovered as $\text{expm}(\widehat{\xi})$ or `jaxlie.SE3.exp(vector)`.

Since the exponential coordinates can take any real numbers, standard ADAM with parameters of shape `(6,)` and gradients of shape `(6,)` can be run without thinking about projection. If you follow NeRF papers this is loosely what iNeRF, BARF, etc do.

**Method 1 (tangent-space ADAM)** is basically Riemannian gradient descent, and closer to how SLAM people think about pose optimization ([relevant gtsam blog post](https://gtsam.org/2021/02/23/uncertainties-part2.html)).

It's kind of a hybrid of the first two methods: we track our transform parameters using the global length-`7` vector, but we compute gradients of shape `(6,)`. At training step $k$ we can compute the gradient $g = \nabla_{\xi}\ f(T_{k} e^{\widehat{\xi}})$, apply ADAM scaling rules, and then instead of Euclidean addition apply the update as $T_{k+1} = T_k e^{\widehat{g}}$. This a "right plus" in this [micro Lie theory paper](https://arxiv.org/abs/1812.01537), which is why the operation used in `jaxlie` is `jaxlie.manifold.rplus()`.

> I'm assuming the practical difference between these methods might be about computing derivatives around certain points in the SE3 space, e.g. like around (axis-angle = [0,0,0,0,0,0]). Does your method (method 1) solve such issues, that I think are present in method 3??

This sounds about right, but if you're referring to singularities near the identity that's not quite it. (for singularities we just use Taylor expansions, you can grep the codebase for `jnp.where`)

The main difference is how nicely structured + locally linear the spaces you're computing derivatives in is, which matters a lot for second-order/Newton methods or more naive optimization approaches. If you use a modern first-order optimizer with bells and whistles like momentum and scale-invariance via adaptive step sizes (like ADAM), my personal experience is that the practical difference is small.

**Comment by JonathonLuiten** [comment]
Awesome!

Thanks a lot for taking your time. This answers my questions nicely.

I'll close this now :)

**Comment by nnop** [comment]
Does the _LieTorch_ method belong to the three methods mentioned above?
